### DISCUSSION TITLE
How can I start to contribute?

### DISCUSSION THREAD

**Discussion by sandangel** [noib3/nvim-completion#8] [General]
I can see lots of TODO in the code, but I don't know which one I should pick up first. Do you have any issue that you can assign to me to get more familiar with the code base?

**Comment by noib3** [comment]
Hi, thanks for reaching out. I guess this: https://github.com/noib3/nvim-compleet/blob/08cd0763f7ba2f62835565c045933b5f66780061/src/ui/details/positioning.rs#L78

would be a good place to start. It's a bug with the positioning of the details window (i.e. the window next to the completion menu that shows details about a completion) that causes `nvim-compleet` to panic.

Demo:
![bug](https://user-images.githubusercontent.com/59321248/161089756-45629309-1539-4cdb-b59a-ae5640f574ba.gif)

to reproduce:

- compile from commit 08cd0763f7ba2f62835565c045933b5f66780061;
- set the columns to 173 (altough it's probably possible to reproduce this with a different terminal width);
- open a new vertical split to the right and go to that window;
- type `s`;
- select the first completion;
- type `i`.

The reason is that `api.win_get_position(menu_winid)` (which is just a Rust binding to `vim.api.nvim_win_get_position`) returns a number of columns bigger than the total terminal width, causing the following unchecked subtractions to overflow.

I believe it's a bug with Neovim itself rather than `nvim-compleet`, but I haven't looked too much into it.

P.S.: this is just a suggestion, if you'd like to work on something else feel free to do so. Also, if you'd like to contact me send me a DM on Telegram (https://t.me/noib3) or on matrix (@noib3:matrix.org).

**Comment by sandangel** [comment]
thank you. I will put some time to investigate that one this weekend.

**Comment by sandangel** [comment]
hmm, I can not reproduce that. set columns=173 and open new vsplit, but  don't know why my menu is different with you when I hit `<c-n>` to go to the next item. I also tried disabling all plugins but still the same. cc @noib3 

```lua
  use {
    '~/Work/nvim-compleet',
    config = function()
      require('compleet').setup {
        sources = {
          lipsum = {
            enable = true,
          },
        },
      }
    end,
  }

```


https://user-images.githubusercontent.com/22189661/161381740-dcbe7cd7-2d9c-4e12-8624-f27f1eee6f63.mov

**Reply by bryant-the-coder** [reply]
> hmm, I can not reproduce that. set columns=173 and open new vsplit, but don't know why my menu is different with you when I hit `<c-n>` to go to the next item. I also tried disabling all plugins but still the same. cc @noib3
> 
> ```lua
>   use {
>     '~/Work/nvim-compleet',
>     config = function()
>       require('compleet').setup {
>         sources = {
>           lipsum = {
>             enable = true,
>           },
>         },
>       }
>     end,
>   }
> ```
> 
> Screen.Recording.2022-04-02.at.20.48.07.mov

what is your dots?

**Comment by noib3** [comment]
I think you haven't set the [mappings](https://github.com/noib3/nvim-compleet#musical_keyboard-mappings) to select the first completion. If that still doesn't work, try to replicate it by creating a minimal `init.lua` and calling neovim with `nvim --clean -u ./init.lua`.

**Comment by nxmxsxs** [comment]
Hi, I noticed you've been working on the back2rpc branch recently. Is that going to be the final architecture? Are you going to merge that eventually into master?

**Reply by noib3** [reply]
The short answer is yes, but please open a new discussion and we can talk about it.

**Reply by nxmxsxs** [reply]
I don't have much experience with rust but I would like to, as a way to learn more about neovim plugins ecosystem/git, contribute in any way possible but I'm exactly sure what I would work on. Could you point me in the right direction?

**Reply by noib3** [reply]
Well this issue would be a good place to start. Also, if you'd like to learn more about both Rust and neovim then look at the `compleet-client` crate of the `back2rpc` branch. The server is what I'm currently working on but that's got nothing to do with neovim per se.

**Comment by sandangel** [comment]
hi @noib3 , I still can not reproduce. the only issue is the menu window disappeared, and the preview window keep showing.

https://user-images.githubusercontent.com/22189661/161424353-2e1efc92-3df7-4f65-acfb-22c6e8b76669.mov




https://user-images.githubusercontent.com/22189661/161424255-c8aba320-c7e2-4b10-97b8-9df7412da408.mov

**Reply by noib3** [reply]
Have you tried with `nvim --clean -u ./init.lua`?

**Reply by sandangel** [reply]
yes, I also tried with `nvim --clean -u ./init.lua`

```lua
local install_path = vim.fn.stdpath 'data' .. '/site/pack/packer/start/packer.nvim'

if vim.fn.empty(vim.fn.glob(install_path)) > 0 then
  vim.fn.system {
    'git',
    'clone',
    '--depth',
    '1',
    'https://github.com/wbthomason/packer.nvim',
    install_path,
  }
end

vim.o.runtimepath = vim.fn.stdpath 'data' .. '/site/pack/*/start/*,' .. vim.o.runtimepath

require('packer').startup(function(use)
  use 'wbthomason/packer.nvim'

  use {
    'navarasu/onedark.nvim',
    config = function()
      require('onedark').setup {
        style = 'dark',
        code_style = {
          comments = 'italic',
          keywords = 'italic',
          functions = 'italic',
          strings = 'none',
          variables = 'italic',
        },
      }
      require('onedark').load()
    end,
  }

  use {
    '~/Work/nvim-compleet',
    config = function()
      local compleet = require 'compleet'
      compleet.setup {
        sources = {
          lipsum = {
            enable = true,
          },
        },
      }
      local keymap = vim.keymap

      local tab = function()
        return (compleet.is_menu_visible() and '<Plug>(compleet-next-completion)')
          or (compleet.has_completions() and '<Plug>(compleet-show-completions)')
          or '<Tab>'
      end

      local s_tab = function()
        return compleet.is_menu_visible() and '<Plug>(compleet-prev-completion)' or '<S-Tab>'
      end

      local right = function()
        return compleet.is_hint_visible() and '<Plug>(compleet-insert-hinted-completion)' or '<Right>'
      end

      local cr = function()
        return compleet.is_completion_selected() and '<Plug>(compleet-insert-selected-completion)' or '<CR>'
      end

      local opts = { expr = true, remap = true }

      keymap.set('i', '<Tab>', tab, opts)
      keymap.set('i', '<S-Tab>', s_tab, opts)
      keymap.set('i', '<Right>', right, opts)
      keymap.set('i', '<CR>', cr, opts)
    end,
  }
end)

```

https://user-images.githubusercontent.com/22189661/161434306-8a859a92-54ce-47b6-a1a0-f043d07fc9f4.mov

**Reply by noib3** [reply]
The right split is too wide. Make sure that `:set columns?` returns `173`, then put the cursor on the right window and make sure that `:lua =vim.api.nvim_win_get_position(0)[2]` returns `87`.

**Reply by nxmxsxs** [reply]
I think I found another bug which signcolumn option which could be related

https://user-images.githubusercontent.com/95099980/161440167-25633ee4-0fa2-4b0b-84ec-9888a2f65b27.mov

.

**Reply by noib3** [reply]
@MurdeRM3L0DY can you open a new issue showing that video?

**Reply by sandangel** [reply]
Hi @noib3 , I already did. Just recorded another video to show you. I only have the minimal config above.
https://user-images.githubusercontent.com/22189661/161467496-63eff412-a6e3-4249-bd16-16dc73431c25.mp4

**Reply by noib3** [reply]
@sandangel `set columns=173` doesn't actually do anything, you have to resize the terminal manually. In your video the right split is much wider than the left one, which has the correct width. Make sure **the output** of `set columns?` is `173`, and **not** `set columns=173`.

**Reply by sandangel** [reply]
@noib3 I tried again after resizing terminal, still can not reproduce.

https://user-images.githubusercontent.com/22189661/161538384-7066655b-7f05-4ff8-aeec-63c7527350d4.mp4

**Reply by noib3** [reply]
I see, I'm not sure why it can't be reproduced. I'll look into it.

**Reply by nxmxsxs** [reply]
That's strange. I'm able to reproduce it on my machine

**Reply by nxmxsxs** [reply]
Hello, I noticed you don't 'vim.schedule' the win_set_config function. I was messing around in lua with floating windows and cursormoved autocommands and I noticed the floating windows don't follow the cursor as you would expect.

https://user-images.githubusercontent.com/95099980/161842389-ae2c42c8-0a4e-4ad5-9702-1e8daee007c1.mov

**Reply by nxmxsxs** [reply]
nvm, you don't need to vim.schedule `CursorMovedI` autocmds.

https://user-images.githubusercontent.com/95099980/161938480-77813bcf-cd85-4306-8a66-399083c7da50.mov

**Reply by noib3** [reply]
> Hello, I noticed you don't 'vim.schedule' the win_set_config function. I was messing around in lua with floating windows and cursormoved autocommands and I noticed the floating windows don't follow the cursor as you would expect.
> 
>  2022-04-05.22-20-06.mov

Mh, I wonder if [this](https://github.com/neovim/neovim/issues/17853) might be related.

**Reply by nxmxsxs** [reply]
possibly. I assumed it was just neovim's single thread scheduling shenanigans.

**Comment by nxmxsxs** [comment]
Hi, looking through the codebase, I can understand marginally how the client comunicates with the server via rpc but I can't figure out how the server would produce lsp completions (in particular) for the client.

**Reply by noib3** [reply]
Well right now it doesn't. The plan is to let the server spawn its own LSP process and send the completions it gets from the LSP to the client

**Reply by nxmxsxs** [reply]
I may be misunderstanding but wouldn't that mean spawning 2 separate LSP processes?

**Reply by noib3** [reply]
Yep, it would. Getting the LSP results directly from Neovim would mean converting possibly hundreds of Lua tables to Rust structs on every keystroke, which is not ideal. Also, the LSP source would need to be treated differently from all the others as it'd be the only one coming from the client, while all the other come from the server.

**Reply by nxmxsxs** [reply]
ah I see. How would compleet-server figure out which LSP server to spawn?
Also both servers would need to comunicate via json-rpc right?

**Reply by noib3** [reply]
The `Source` trait contains an `attach` function that the source can use to figure out if it should attach to the buffer. In the case of LSP it would ask the client to send the output of `vim.lsp.buf_get_clients()`. Yes, LSPs communicate using json-rpc.

**Reply by nxmxsxs** [reply]
I was referring to the comunication between compleet-server and LSP server? Would we have to implement json-rpc for compleet-server?

**Reply by noib3** [reply]
Yep, shouldn't bee to hard using serde.

**Reply by sandangel** [reply]
I'm trying to compile the project with latest master, but `cargo build` failed with error:

```sh
= note: Undefined symbols for architecture x86_64:
```

I'm using rust nightly 1.62 on mac os 2019

**Reply by noib3** [reply]
@sandangel Mh, does it compile if you move `/compleet-client/.cargo` to `/.cargo`?

**Reply by nxmxsxs** [reply]
> Yep, shouldn't bee to hard using serde.

Maybe, we could use [jsonrpsee](https://github.com/paritytech/jsonrpsee)?

**Reply by sandangel** [reply]
> @sandangel Mh, does it compile if you move `/compleet-client/.cargo` to `/.cargo`?

yeah it does compile now.

**Comment by nxmxsxs** [comment]
Hi, would you happen to know of a way to check if a win is split (vertical or horizontal).
I've tried with `win_height <= lines/2 or win_width <= columns/2` but this doesn't cover all edge cases.

**Reply by noib3** [reply]
Look into `vim.fn.winlayout`. I'm using it in `nvim-cokeline` [here](https://github.com/noib3/nvim-bufferline/blob/e852df1b6d5ce57a871d13d6da22e59968e26561/lua/cokeline/sidebar.lua#L21) to check for vertical splits.

**Reply by nxmxsxs** [reply]
thanks. I'll look into it.

**Comment by nxmxsxs** [comment]
Hello, I'm working on a branch which changes the available screen space calculation but upon testing, compleet panics when I enable a source. Is it possible that this behaviour is from your part?

**Reply by noib3** [reply]
It could be. When does it panic and what's the error message?

**Reply by nxmxsxs** [reply]
the error message is `attempt to index local 'error_msg' (a userdata value)` and it happens when I enable any source.

**Reply by noib3** [reply]
But is it a Lua error or a Rust panic? If it's a panic, what's the line that caused it? Also, make sure to update your branch by pulling in master. A lot changed in the past week.

**Comment by nxmxsxs** [comment]
The issue only happens when I call setup and enable one of the sources so I assume it's a rust panic.
I've also tried cloning your master but same result.

https://user-images.githubusercontent.com/95099980/163577442-02d00c26-8918-4e47-aafc-1106deddf201.mp4

**Reply by noib3** [reply]
No that's a Lua runtime error, if it was a panic Neovim would crash. What's your `nvim --version`? I can't reproduce it. Also what's the output of `:lua =vim.loop.new_pipe`?

**Reply by nxmxsxs** [reply]
I'm on latest master (0.8.0-dev).
`vim.loop.new_pipe` exists and is a function as expected

**Reply by noib3** [reply]
Mh, that's weird. Try commenting out this whole block:

https://github.com/noib3/nvim-compleet/blob/063ca8932a834bc8cdbbf0db6a7b0e027f4aaa26/compleet-client/src/setup.rs#L99-L110

If that works start uncommenting code until you find the line that's causing the error.

**Reply by nxmxsxs** [reply]
Yeah, that disables the plugin as expected and eliminates the error message as well

**Reply by nxmxsxs** [reply]
It seems the error message originates from `command::setup`

**Reply by noib3** [reply]
Ah, then I'm pretty sure one of the functions exposed by Neovim to set user commands changed its api. When did you start experiencing this? Does it still happen if you use a Neovim version compiled from one of last week's commits?

**Reply by nxmxsxs** [reply]
oh right, I forgor they changed `nvim_add_user_command` to `nvim_create_user_command`. Yeah that fixes it

**Reply by noib3** [reply]
Cool, if you send a PR I'll merge it right away. You need to update these lines

https://github.com/noib3/nvim-compleet/blob/063ca8932a834bc8cdbbf0db6a7b0e027f4aaa26/compleet-client/src/commands/setup.rs#L39-L40

https://github.com/noib3/nvim-compleet/blob/063ca8932a834bc8cdbbf0db6a7b0e027f4aaa26/compleet-client/src/bindings/api/global.rs#L8-L17

**Comment by nxmxsxs** [comment]
Had [this](https://github.com/noib3/nvim-compleet/discussions/8#discussioncomment-2492691) already been fixed in master?

https://user-images.githubusercontent.com/95099980/163629574-95260229-ab32-436f-9c04-5f4a23a970a5.mp4

**Reply by noib3** [reply]
nope

**Comment by nxmxsxs** [comment]
Hello, What would you think about all the crates into a single `compleet` lib crate?

**Reply by noib3** [reply]
What would be the benefit of that? Having separate crates allows for separation of concerns and reduced build times.

**Reply by nxmxsxs** [reply]
Hmm, I didn't know about the build time improvement. How do you plan on letting users add new sources via rust? I was thinking maybe users would be able to add compleet to the dependencies list with only `sources` feature or something similar.

**Reply by noib3** [reply]
All the sources are in the `compleet-sources` crate. If contributors want to add new Rust sources they'll send a PR.

**Reply by sandangel** [reply]
I dont think it is a scalable approach. you will introduce a bottle neck and create a burden for yourself reviewing all the code. The best way is you publish the interface and all devs can create their own source by using that interface.

**Reply by noib3** [reply]
@sandangel what I've described is the process for sources everyone would want to use, things like lsp, filepaths, snippets, etc. Those are part of the core plugin. Scalability shouldn't be a problem, there's a limited amount of such "global" sources. If users want to add a custom source that's only useful to them they can just add it to their copy of the source w/o sending a PR.

**Reply by sandangel** [reply]
oh, that makes sense now. 😊

**Comment by sandangel** [comment]
@noib3 as I check the source code, it seems like we need to bind any neovim API we need to call in the rust code by adding an enum value in `BridgeRequest`, and a new method in `impl Neovim`. Is there a more generic way to do that? Like how we call `vim.cmd` in lua code, the `vim.cmd` can run any arbitrary vim command. I think the current approach will need us to bind every lua function/method we need to use, which is a lot of work and hard to maintain 😣

**Reply by sandangel** [reply]
I found https://github.com/little-dude/rmp-rpc , a message pack rpc implementation in rust. Message Pack RPC is used by neovim for programmatically control neovim and listen to its events. I think we can use that instead

**Reply by noib3** [reply]
> it seems like we need to bind any neovim API we need to call in the rust code by adding an enum value in `BridgeRequest`, and a new method in `impl Neovim`. Is there a more generic way to do that?

That's correct. I don't see how to avoid this, every request has different parameters, different return types and different ways to handle it (e.g. `LpsBufGetClients` and `LspClientRequest` are handled way differently than `ApiBufGetName` and `ApiGetCurrentBuf`.

> I found https://github.com/little-dude/rmp-rpc , a message pack rpc implementation in rust. Message Pack RPC is used by neovim for programmatically control neovim and listen to its events. I think we can use that instead

That would require `nvim-compleet` to be spawned as an external process and communicate with the client via rpc. I tried that approach already but ultimately decided against it: implementing an rpc server and serializing/deserializing messages introduces way more complexity that adding a few enum variants to `BridgeRequest` as they become necessary.

**Comment by marvindore** [comment]
Hello thank you for creating nvim_completion it seems very interesting. I am currently teaching myself Rust so I am lacking in Rust specific knowledge. 
Are there instructions on how to get this plugin to work?

I forked the repo and loaded it using Packer.nvim but after running my configuration file it says cannot find module nvim_completion. Any setup steps to get this running is appreciated. Below is my config.

use { '~/home/nvim-completion', branch = 'master' }

require('nvim_completion').setup {
        sources = {
          lipsum = { enable = function(_buf) return true end },
          lsp = { enable = true },
        },
      }
